### DISCUSSION TITLE
Upgrade to 4.0 loses some variables from 3.0 dump

### DISCUSSION THREAD

**Discussion by blind-oracle** [iNavFlight/inav#7685] [Q&A] [answered]
Just upgraded to 4.0:
1. Did a `diff all` in 3.0
2. Upgraded configurator to 4.0
3. Flashed with full erase
4. Restored dump through CLI tab

At least `throttle_idle` and `nav_mc_hover_thr` settings left with default values instead of the ones from the dump.
Though during dump loading the CLI definitely said that these variables were set to correct values, but after reboot they're default.

Maybe some others also got lost, haven't checked yet.

**Comment by stronnag** [comment, marked as answer]
Check the battery profiles. At least `throttle_idle` is profile dependent (and it works for me once I evinced that).

**Comment by blind-oracle** [comment]
@stronnag Thanks, that wasn't obvious, especially the idle throttle. But after digging in sources I see the relevant struct: https://github.com/iNavFlight/inav/blob/7dad0bfa681d7940f49ec9932be924b2b6f09aa9/src/main/sensors/battery_config_structs.h#L77

I guess we need some docs update, otherwise it might surprise others. Will do a PR maybe.

**Reply by stronnag** [reply]
> I guess we need some docs update, otherwise it might surprise others. Will do a PR maybe.

It would certainly be useful to indicate whether a value is global, profile or battery profile specific.

**Comment by stronnag** [comment]
Not an inconsiderable list ...

## Battery Profile

| Setting | Description |
| ------- | ----------- |
| bat_cells | Number of cells of the battery (0 = auto-detect), see battery documentation. 7S, 9S and 11S batteries cannot be auto-detected. |
| vbat_cell_detect_voltage | Maximum voltage per cell, used for auto-detecting the number of cells of the battery in 0.01V units. |
| vbat_max_cell_voltage | Maximum voltage per cell in 0.01V units, default is 4.20V |
| vbat_min_cell_voltage | Minimum voltage per cell, this triggers battery out alarms, in 0.01V units, default is 330 (3.3V) |
| vbat_warning_cell_voltage | Warning voltage per cell, this triggers battery-warning alarms, in 0.01V units, default is 350 (3.5V) |
| battery_capacity | Set the battery capacity in mAh or mWh (see `battery_capacity_unit`). Used to calculate the remaining battery capacity. |
| battery_capacity_warning | If the remaining battery capacity goes below this threshold the beeper will emit short beeps and the relevant OSD items will blink. |
| battery_capacity_critical | If the remaining battery capacity goes below this threshold the battery is considered empty and the beeper will emit long beeps. |
| battery_capacity_unit | Unit used for `battery_capacity`, `battery_capacity_warning` and `battery_capacity_critical` [MAH/MWH] (milliAmpere hour / milliWatt hour). |
| controlrate_profile | Control rate profile to switch to when the battery profile is selected, 0 to disable and keep the currently selected control rate profile |
| throttle_scale | Throttle scaling factor. `1` means no throttle scaling. `0.5` means throttle scaled down by 50% |
| throttle_idle | The percentage of the throttle range (`max_throttle` - `min_command`) above `min_command` used for minimum / idle throttle. |
| turtle_mode_power_factor | Turtle mode power factor |
| failsafe_throttle | Throttle level used for landing when failsafe is enabled. See [Failsafe documentation](Failsafe.md#failsafe_throttle). |
| fw_min_throttle_down_pitch | Automatic pitch down angle when throttle is at 0 in angle mode. Progressively applied between cruise throttle and zero throttle (decidegrees) |
| nav_mc_hover_thr | Multicopter hover throttle hint for altitude controller. Should be set to approximate throttle value when drone is hovering. |
| nav_fw_cruise_thr | Cruise throttle in GPS assisted modes, this includes RTH. Should be set high enough to avoid stalling. This values gives INAV a base for throttle when flying straight, and it will increase or decrease throttle based on pitch of airplane and the parameters below. In addition it will increase throttle if GPS speed gets below 7m/s ( hardcoded ) |
| nav_fw_min_thr | Minimum throttle for flying wing in GPS assisted modes |
| nav_fw_max_thr | Maximum throttle for flying wing in GPS assisted modes |
| nav_fw_pitch2thr | Amount of throttle applied related to pitch attitude in GPS assisted modes. Throttle = nav_fw_cruise_throttle - (nav_fw_pitch2thr * pitch_angle). (notice that pitch_angle is in degrees and is negative when climbing and positive when diving, and throttle value is constrained between nav_fw_min_thr and nav_fw_max_thr) |
| nav_fw_launch_thr | Launch throttle - throttle to be set during launch sequence (pwm units) |
| nav_fw_launch_idle_thr | Launch idle throttle - throttle to be set before launch sequence is initiated. If set below minimum throttle it will force motor stop or at idle throttle (depending if the MOTOR_STOP is enabled). If set above minimum throttle it will force throttle to this value (if MOTOR_STOP is enabled it will be handled according to throttle stick position) |
| limit_cont_current | Continous current limit (dA), set to 0 to disable |
| limit_burst_current | Burst current limit (dA): the current which is allowed during `limit_burst_current_time` after which `limit_cont_current` will be enforced, set to 0 to disable |
| limit_burst_current_time | Allowed current burst time (ds) during which `limit_burst_current` is allowed and after which `limit_cont_current` will be enforced |
| limit_burst_current_falldown_time | Time slice at the end of the burst time during which the current limit will be ramped down from `limit_burst_current` back down to `limit_cont_current` |
| limit_cont_power | Continous power limit (dW), set to 0 to disable |
| limit_burst_power | Burst power limit (dW): the current which is allowed during `limit_burst_power_time` after which `limit_cont_power` will be enforced, set to 0 to disable |
| limit_burst_power_time | Allowed power burst time (ds) during which `limit_burst_power` is allowed and after which `limit_cont_power` will be enforced |
| limit_burst_power_falldown_time | Time slice at the end of the burst time during which the power limit will be ramped down from `limit_burst_power` back down to `limit_cont_power` |
## Profile - Control Rate

| Setting | Description |
| ------- | ----------- |
| thr_mid | Throttle value when the stick is set to mid-position. Used in the throttle curve calculation. |
| thr_expo | Throttle exposition value |
| tpa_rate | Throttle PID attenuation reduces influence of P on ROLL and PITCH as throttle increases. For every 1% throttle after the TPA breakpoint, P is reduced by the TPA rate. |
| tpa_breakpoint | See tpa_rate. |
| fw_tpa_time_constant | TPA smoothing and delay time constant to reflect non-instant speed/throttle response of the plane. Planes with low thrust/weight ratio generally need higher time constant. Default is zero for compatibility with old setups |
| rc_expo | Exposition value used for the PITCH/ROLL axes by all the stabilized flights modes (all but `MANUAL`) |
| rc_yaw_expo | Exposition value used for the YAW axis by all the stabilized flights modes (all but `MANUAL`) |
| roll_rate | Defines rotation rate on ROLL axis that UAV will try to archive on max. stick deflection. Rates are defined in tens of degrees (deca-degrees) per second [rate = dps/10]. That means, rate 20 represents 200dps rotation speed. Default 20 (200dps) is more less equivalent of old Cleanflight/Baseflight rate 0. Max. 180 (1800dps) is what gyro can measure. |
| pitch_rate | Defines rotation rate on PITCH axis that UAV will try to archive on max. stick deflection. Rates are defined in tens of degrees (deca-degrees) per second [rate = dps/10]. That means, rate 20 represents 200dps rotation speed. Default 20 (200dps) is more less equivalent of old Cleanflight/Baseflight rate 0. Max. 180 (1800dps) is what gyro can measure. |
| yaw_rate | Defines rotation rate on YAW axis that UAV will try to archive on max. stick deflection. Rates are defined in tens of degrees (deca-degrees) per second [rate = dps/10]. That means, rate 20 represents 200dps rotation speed. Default 20 (200dps) is more less equivalent of old Cleanflight/Baseflight rate 0. Max. 180 (1800dps) is what gyro can measure. |
| manual_rc_expo | Exposition value used for the PITCH/ROLL axes by the `MANUAL` flight mode [0-100] |
| manual_rc_yaw_expo | Exposition value used for the YAW axis by the `MANUAL` flight mode [0-100] |
| manual_roll_rate | Servo travel multiplier for the ROLL axis in `MANUAL` flight mode [0-100]% |
| manual_pitch_rate | Servo travel multiplier for the PITCH axis in `MANUAL` flight mode [0-100]% |
| manual_yaw_rate | Servo travel multiplier for the YAW axis in `MANUAL` flight mode [0-100]% |
| fpv_mix_degrees |  |
| rate_dynamics_center_sensitivity | The center stick sensitivity for Rate Dynamics |
| rate_dynamics_end_sensitivity | The end stick sensitivity for Rate Dynamics |
| rate_dynamics_center_correction | The center stick correction for Rate Dynamics |
| rate_dynamics_end_correction | The end  stick correction for Rate Dynamics |
| rate_dynamics_center_weight | The center stick weight for Rate Dynamics |
| rate_dynamics_end_weight | The end  stick weight for Rate Dynamics |
## Profile - PID

| Setting | Description |
| ------- | ----------- |
| mc_p_pitch | Multicopter rate stabilisation P-gain for PITCH |
| mc_i_pitch | Multicopter rate stabilisation I-gain for PITCH |
| mc_d_pitch | Multicopter rate stabilisation D-gain for PITCH |
| mc_cd_pitch | Multicopter Control Derivative gain for PITCH |
| mc_p_roll | Multicopter rate stabilisation P-gain for ROLL |
| mc_i_roll | Multicopter rate stabilisation I-gain for ROLL |
| mc_d_roll | Multicopter rate stabilisation D-gain for ROLL |
| mc_cd_roll | Multicopter Control Derivative gain for ROLL |
| mc_p_yaw | Multicopter rate stabilisation P-gain for YAW |
| mc_i_yaw | Multicopter rate stabilisation I-gain for YAW |
| mc_d_yaw | Multicopter rate stabilisation D-gain for YAW |
| mc_cd_yaw | Multicopter Control Derivative gain for YAW |
| mc_p_level | Multicopter attitude stabilisation P-gain |
| mc_i_level | Multicopter attitude stabilisation low-pass filter cutoff |
| mc_d_level | Multicopter attitude stabilisation HORIZON transition point |
| fw_p_pitch | Fixed-wing rate stabilisation P-gain for PITCH |
| fw_i_pitch | Fixed-wing rate stabilisation I-gain for PITCH |
| fw_d_pitch | Fixed wing rate stabilisation D-gain for PITCH |
| fw_ff_pitch | Fixed-wing rate stabilisation FF-gain for PITCH |
| fw_p_roll | Fixed-wing rate stabilisation P-gain for ROLL |
| fw_i_roll | Fixed-wing rate stabilisation I-gain for ROLL |
| fw_d_roll | Fixed wing rate stabilisation D-gain for ROLL |
| fw_ff_roll | Fixed-wing rate stabilisation FF-gain for ROLL |
| fw_p_yaw | Fixed-wing rate stabilisation P-gain for YAW |
| fw_i_yaw | Fixed-wing rate stabilisation I-gain for YAW |
| fw_d_yaw | Fixed wing rate stabilisation D-gain for YAW |
| fw_ff_yaw | Fixed-wing rate stabilisation FF-gain for YAW |
| fw_p_level | Fixed-wing attitude stabilisation P-gain |
| fw_i_level | Fixed-wing attitude stabilisation low-pass filter cutoff |
| fw_d_level | Fixed-wing attitude stabilisation HORIZON transition point |
| max_angle_inclination_rll | Maximum inclination in level (angle) mode (ROLL axis). 100=10° |
| max_angle_inclination_pit | Maximum inclination in level (angle) mode (PITCH axis). 100=10° |
| dterm_lpf_hz | Dterm low pass filter cutoff frequency. Default setting is very conservative and small multirotors should use higher value between 80 and 100Hz. 80 seems like a gold spot for 7-inch builds while 100 should work best with 5-inch machines. If motors are getting too hot, lower the value |
| dterm_lpf_type | Defines the type of stage 1 D-term LPF filter. Possible values: `PT1`, `BIQUAD`, `PT2`, `PT3`. |
| dterm_lpf2_hz | Cutoff frequency for stage 2 D-term low pass filter |
| dterm_lpf2_type | Defines the type of stage 1 D-term LPF filter. Possible values: `PT1`, `BIQUAD`, `PT2`, `PT3`. |
| yaw_lpf_hz | Yaw P term low pass filter cutoff frequency. Should be disabled (set to `0`) on small multirotors (7 inches and below) |
| fw_iterm_throw_limit | Limits max/min I-term value in stabilization PID controller in case of Fixed Wing. It solves the problem of servo saturation before take-off/throwing the airplane into the air. By default, error accumulated in I-term can not exceed 1/3 of servo throw (around 165us). Set 0 to disable completely. |
| fw_loiter_direction | Direction of loitering: center point on right wing (clockwise - default), or center point on left wing (counterclockwise). If equal YAW then can be changed in flight using a yaw stick. |
| fw_reference_airspeed | Reference airspeed. Set this to airspeed at which PIDs were tuned. Usually should be set to cruise airspeed. Also used for coordinated turn calculation if airspeed sensor is not present. |
| fw_turn_assist_yaw_gain | Gain required to keep the yaw rate consistent with the turn rate for a coordinated turn (in TURN_ASSIST mode). Value significantly different from 1.0 indicates a problem with the airspeed calibration (if present) or value of `fw_reference_airspeed` parameter |
| fw_turn_assist_pitch_gain | Gain required to keep constant pitch angle during coordinated turns (in TURN_ASSIST mode). Value significantly different from 1.0 indicates a problem with the airspeed calibration (if present) or value of `fw_reference_airspeed` parameter |
| fw_iterm_limit_stick_position | Iterm is not allowed to grow when stick position is above threshold. This solves the problem of bounceback or followthrough when full stick deflection is applied on poorely tuned fixed wings. In other words, stabilization is partialy disabled when pilot is actively controlling the aircraft and active when sticks are not touched. `0` mean stick is in center position, `1` means it is fully deflected to either side |
| fw_yaw_iterm_freeze_bank_angle | Yaw Iterm is frozen when bank angle is above this threshold [degrees]. This solves the problem of the rudder counteracting turns by partially disabling yaw stabilization when making banked turns. Setting to 0 (the default) disables this feature. Only applies when autopilot is not active and TURN ASSIST is disabled. |
| pidsum_limit | A limitation to overall amount of correction Flight PID can request on each axis (Roll/Pitch). If when doing a hard maneuver on one axis machine looses orientation on other axis - reducing this parameter may help |
| pidsum_limit_yaw | A limitation to overall amount of correction Flight PID can request on each axis (Yaw). If when doing a hard maneuver on one axis machine looses orientation on other axis - reducing this parameter may help |
| iterm_windup | Used to prevent Iterm accumulation on during maneuvers. Iterm will be dampened when motors are reaching it's limit (when requested motor correction range is above percentage specified by this parameter) |
| rate_accel_limit_roll_pitch | Limits acceleration of ROLL/PITCH rotation speed that can be requested by stick input. In degrees-per-second-squared. Small and powerful UAV flies great with high acceleration limit ( > 5000 dps^2 and even > 10000 dps^2). Big and heavy multirotors will benefit from low acceleration limit (~ 360 dps^2). When set correctly, it greatly improves stopping performance. Value of 0 disables limiting. |
| rate_accel_limit_yaw | Limits acceleration of YAW rotation speed that can be requested by stick input. In degrees-per-second-squared. Small and powerful UAV flies great with high acceleration limit ( > 10000 dps^2). Big and heavy multirotors will benefit from low acceleration limit (~ 180 dps^2). When set correctly, it greatly improves stopping performance and general stability during yaw turns. Value of 0 disables limiting. |
| heading_hold_rate_limit | This setting limits yaw rotation rate that HEADING_HOLD controller can request from PID inner loop controller. It is independent from manual yaw rate and used only when HEADING_HOLD flight mode is enabled by pilot, RTH or WAYPOINT modes. |
| nav_mc_pos_z_p | P gain of altitude PID controller (Multirotor) |
| nav_mc_vel_z_p | P gain of velocity PID controller |
| nav_mc_vel_z_i | I gain of velocity PID controller |
| nav_mc_vel_z_d | D gain of velocity PID controller |
| nav_mc_pos_xy_p | Controls how fast the drone will fly towards the target position. This is a multiplier to convert displacement to target velocity |
| nav_mc_vel_xy_p | P gain of Position-Rate (Velocity to Acceleration) PID controller. Higher P means stronger response when position error occurs. Too much P might cause "nervous" behavior and oscillations |
| nav_mc_vel_xy_i | I gain of Position-Rate (Velocity to Acceleration) PID controller. Used for drift compensation (caused by wind for example). Higher I means stronger response to drift. Too much I gain might cause target overshot |
| nav_mc_vel_xy_d | D gain of Position-Rate (Velocity to Acceleration) PID controller. It can damp P and I. Increasing D might help when drone overshoots target. |
| nav_mc_vel_xy_ff |  |
| nav_mc_heading_p | P gain of Heading Hold controller (Multirotor) |
| nav_mc_vel_xy_dterm_lpf_hz |  |
| nav_mc_vel_xy_dterm_attenuation | Maximum D-term attenution percentage for horizontal velocity PID controller (Multirotor). It allows to smooth the PosHold CRUISE, WP and RTH when Multirotor is traveling at full speed. Dterm is not attenuated at low speeds, breaking and accelerating. |
| nav_mc_vel_xy_dterm_attenuation_start | A point (in percent of both target and current horizontal velocity) where nav_mc_vel_xy_dterm_attenuation begins |
| nav_mc_vel_xy_dterm_attenuation_end | A point (in percent of both target and current horizontal velocity) where nav_mc_vel_xy_dterm_attenuation reaches maximum |
| nav_fw_pos_z_p | P gain of altitude PID controller (Fixedwing) |
| nav_fw_pos_z_i | I gain of altitude PID controller (Fixedwing) |
| nav_fw_pos_z_d | D gain of altitude PID controller (Fixedwing) |
| nav_fw_pos_xy_p | P gain of 2D trajectory PID controller. Play with this to get a straight line between waypoints or a straight RTH |
| nav_fw_pos_xy_i | I gain of 2D trajectory PID controller. Too high and there will be overshoot in trajectory. Better start tuning with zero |
| nav_fw_pos_xy_d | D gain of 2D trajectory PID controller. Too high and there will be overshoot in trajectory. Better start tuning with zero |
| nav_fw_heading_p | P gain of Heading Hold controller (Fixedwing) |
| nav_fw_pos_hdg_p | P gain of heading PID controller. (Fixedwing, rovers, boats) |
| nav_fw_pos_hdg_i | I gain of heading trajectory PID controller. (Fixedwing, rovers, boats) |
| nav_fw_pos_hdg_d | D gain of heading trajectory PID controller. (Fixedwing, rovers, boats) |
| nav_fw_pos_hdg_pidsum_limit | Output limit for heading trajectory PID controller. (Fixedwing, rovers, boats) |
| mc_iterm_relax |  |
| mc_iterm_relax_cutoff |  |
| d_boost_min |  |
| d_boost_max |  |
| d_boost_max_at_acceleration |  |
| d_boost_gyro_delta_lpf_hz |  |
| antigravity_gain | Max Antigravity gain. `1` means Antigravity is disabled, `2` means Iterm is allowed to double during rapid throttle movements |
| antigravity_accelerator |  |
| antigravity_cutoff_lpf_hz | Antigravity cutoff frequenct for Throtte filter. Antigravity is based on the difference between actual and filtered throttle input. The bigger is the difference, the bigger Antigravity gain |
| pid_type | Allows to set type of PID controller used in control loop. Possible values: `NONE`, `PID`, `PIFF`, `AUTO`. Change only in case of experimental platforms like VTOL, tailsitters, rovers, boats, etc. Airplanes should always use `PIFF` and multirotors `PID` |
| mc_cd_lpf_hz | Cutoff frequency for Control Derivative. Lower value smoother reaction on fast stick movements. With higher values, response will be more aggressive, jerky |
| fw_level_pitch_trim | Pitch trim for self-leveling flight modes. In degrees. +5 means airplane nose should be raised 5 deg from level |
| smith_predictor_strength | The strength factor of a Smith Predictor of PID measurement. In percents |
| smith_predictor_delay | Expected delay of the gyro signal. In milliseconds |
| smith_predictor_lpf_hz | Cutoff frequency for the Smith Predictor Low Pass Filter |
| fw_level_pitch_gain | I-gain for the pitch trim for self-leveling flight modes. Higher values means that AUTOTRIM will be faster but might introduce oscillations |
